optimize/global.go

```diff
@@ -184,158 +177,169 @@ func minimizeGlobal(p *Problem, method GlobalMethod, settings *Settings, stats *
+	workerChan := make(chan GlobalTask)         // Delegate tasks to the workers.
+	evalStatsChan := make(chan GlobalTask)      // Send evaluation updates.
+	statusSent := make(chan struct{})           // Communicate the optimization is done.
 
-// getStatus returns the current status of the optimization.
-func (g *globalStatus) getStatus() Status {
-	var status Status
-	g.mux.RLock()
-	defer g.mux.RUnlock()
-	status = g.status
-	return status
-}
-
-func (g *globalStatus) incrementMajorIteration() {
-	g.mux.Lock()
-	defer g.mux.Unlock()
-	g.stats.MajorIterations++
-}
-
-func (g *globalStatus) updateOptLoc(loc *Location) {
-	g.mux.Lock()
-	defer g.mux.Unlock()
-	copyLocation(g.optLoc, loc)
-}
-
-// checkConvergence checks the convergence of the global optimization and returns
-// the status
-func (g *globalStatus) checkConvergence() Status {
-	g.mux.RLock()
-	defer g.mux.RUnlock()
-	return checkConvergence(g.optLoc, g.settings, false)
-}
-
-// updateStats updates the evaluation statistics for the given operation.
-func (g *globalStatus) updateStats(op Operation) {
-	g.mux.Lock()
-	defer g.mux.Unlock()
-	updateEvaluationStats(g.stats, op)
-}
+	// Launch the method.
+	go func() {
+		tasks := make([]GlobalTask, nTasks)
+		for i := range tasks {
+			tasks[i].Location = newLocation(dim, method)
+		}
+		method.RunGlobal(operations, results, tasks)
+	}()
 
-// updateStatus updates the status and error fields of g. This update only happens
-// if status == NotTerminated, so that the first different status is the one
-// maintained.
-func (g *globalStatus) updateStatus(s Status, err error) {
-	g.mux.Lock()
-	defer g.mux.Unlock()
-	if s != NotTerminated {
-		g.status = s
-		g.err = err
+	// Launch the workers. After workerChan is closed, each worker signals
+	// to the stats updater that they have successfully terminated.
+	for worker := 0; worker < nTasks; worker++ {
+		go func() {
+			x := make([]float64, dim)
+			for task := range workerChan {
+				evaluate(prob, task.Location, task.Operation, x)
+				evalStatsChan <- task
+			}
+			evalStatsChan <- GlobalTask{Operation: signalOperation}
+		}()
 	}
-}
 
-func (g *globalStatus) finishIteration(status Status, err error, loc *Location, op Operation) (Status, error) {
-	g.mux.Lock()
-	defer g.mux.Unlock()
-	return finishIteration(status, err, g.stats, g.settings, g.statuser, g.startTime, loc, op)
-}
+	// Launch the distributor
+	go func() {
+		// Note: This cannot simply be a range loop over operations because we want to
+		// still send to operation even after a termination Status has been sent.
+		for {
+			select {
+			case <-statusSent:
+				close(workerChan)
+				// Read the final operations, only performing an action if MajorIteration.
+				for task := range operations {
+					switch task.Operation {
+					case MajorIteration:
+						evalStatsChan <- task
+					}
+				}
+				// All stats have been sent. Close the channel and wait until all are updated.
```

Comment:
Is this comment still useful, I think the code is clear. At least, there is no wait here anymore.